### PR TITLE
Use Curly instead of Cmd to interact with github

### PR DESCRIPTION
**Pull request by gpetiot**
I need to run some tests and then we could see if we can remove some options and simplify this code.

### CODE REVIEW

**Comment by gpetiot** [comment]
The manual tests on `dune-release-testing` are OK.

**Comment by gpetiot** [comment]
@NathanReb how about I start an intermediate PR where the curl requests are isolated, and we would have `Curl.blabla` functions that return the curl requests (as strings) before they are processed in github.ml ? this way we could add unit tests for the requests, and we can check that `Curly` makes the same requests in this PR

**Comment by NathanReb** [comment]
That sounds like a very good idea! I'm always up for unit tests!

**Comment by gpetiot** [comment]
I will make some tests (but was already tested before externalizing the url/args in Curl.ml), I think it can be reviewed again.

File: lib/github.ml [hunk 1]
```diff
@@ -266,14 +267,12 @@ let assert_tag_exists ~dry_run tag =
...
 let publish_distrib ~dry_run ~msg ~archive ~yes p =
   let git_for_repo r = Cmd.of_list (Cmd.to_list @@ Vcs.cmd r) in
-  let curl = Cmd.(v "curl") in
   ( match Pkg.distrib_user_and_repo p with
   | Error _ as e -> if dry_run then Ok (D.user, D.repo) else e
   | r -> r )
   >>= fun (user, repo) ->
   Pkg.tag p >>= fun tag ->
   assert_tag_exists ~dry_run tag >>= fun () ->
-  OS.Cmd.must_exist curl >>= fun curl ->
```
Comment (NathanReb) [inline lib/github.ml:276]: Is there any particular reason why we removed this check?

I'm not sure how this was reported and how this is reported now but it miight be worth taking a look. What ultimately matters is that we end up with a good error message if `curl` isn't installed. Would mind checking how it was and how it is now?

File: lib/github.ml [hunk 2]
```diff
@@ -175,42 +175,49 @@ let publish_doc ~dry_run ~msg:_ ~docdir ~yes p =
 let github_auth ~dry_run ~user token =
   Sos.read_file ~dry_run token >>= fun token -> Ok (strf "%s:%s" user token)
 
-let run_with_auth ~dry_run auth curl k =
-  let auth = strf "-u %s" auth in
-  Sos.run_io ~dry_run curl (OS.Cmd.in_string auth) k
+let run_with_auth ?(default_body = "") ~dry_run ~auth Curl.{ url; args } =
+  let args = "-u" :: auth :: args in
```
Comment (NathanReb) [inline lib/github.ml:179]: What would you think about extracting a `val with_auth : auth: string -> t -> t` in `Curl` and test it?

File: lib/github.ml (hunk 1 above)
Comment (gpetiot) [reply lib/github.ml:276]: Previously we would get an `Error` here, with this PR, we get an `Error` when calling `curl_create_release` then `run_with_auth` (the Error is returned by `Curly.run`) so the only difference is App_log printed more stuff before getting the error.

File: lib/github.ml (hunk 2 above)
Comment (gpetiot) [reply lib/github.ml]: Good idea, done.

File: lib/github.ml (hunk 1 above)
Comment (NathanReb) [reply lib/github.ml:276]: That's fine by me, do you know what is the error message like when curl is missing?

Comment (gpetiot) [reply lib/github.ml:276]: On master the message was `main.exe: [ERROR] curl: no such command in PATH`, it became:
```
main.exe: [ERROR] curl execution failed: Non 0 exit code Exit code
                  127
                  {status=Exit code 127; stderr="/bin/sh: 1: curl: not
                  found
                  ";
                  stdout=""}
```
the message from Curly is not very clear indeed, I kept the check just before calling `Curly.run`

**Pull request merged**